crypto/evp/p_lib.c

```diff
@@ -1955,12 +1977,22 @@ int evp_pkey_downgrade(EVP_PKEY *pk)
         tmp_copy.keydata = NULL;
 
         evp_pkey_free_it(&tmp_copy);
+        rv = 1;
+    } else {
+        /* Grab the temporary lock to avoid lock leak */
+        tmp_lock = pk->lock;
 
-        return 1;
+        /* Restore the original key */
+        *pk = tmp_copy;          /* |pk| now owns THE lock */
```

Comment:
I must be missing something here...
Doesnt this else mean tmp_lock and pk->lock both point to the same location.. so doing the lock_free(0 further down means we end up with a floating pointer?

Reply:
Ahh I see evp_pkey_reset_unlocked)_ blats it - that is pretty ugly..

I suggest removing the pk == NULL check from the reset_unlocked as coverity is probably going to complain about that, if it gets into this else.

Reply:
@levitte?

Reply:
Er, why would coverity complain about the `pk == NULL` check?  If at some point in the future, we use it somewhere else and happen to pass a NULL, isn't it better that it guards itself, or is it better if it crashes because of NULL dereference?

Reply:
The check in the code implies that it could be NULL..
And then in the :end you use the value..

Reply:
Ah, but an added NULL check in the beginning of *this* function would fix that.

Reply:
Done